### DISCUSSION TITLE
Selection and modification

### DISCUSSION THREAD

**Discussion by morellik** [dflib/dflib#284] [Q&A]
Is it possible to apply modification during a selection?

For example, I've the same loop where I create a new DataFrame with the selected results.
for (List<String> l : entry.getValue()) {
                df1.add(PDBDataFrame.rows(r -> r.get("atomNumber", Integer.class).equals(Integer.parseInt(l.get(8)))).select());
            }

For each row that satisfies the select, if one or two fields in the row satisfy a certain condition, another field changes in value before being added to df1.

**Comment by morellik** [comment]
I tried  mapConditionVal but is deprecated.

**Comment by andrus** [comment]
Yes, you can pass a variety of mapping functions to either `RowSet.select(..)` or `RowSet.map(..)` per https://dflib.org/docs/1.x/#_transform_rows . This way you can transform the existing columns, add new ones, etc.

> I tried mapConditionVal but is deprecated.

If you look at the source of this method, it is deprecated in favor of `Exp.mapBoolVal(..)`. Oh, and I need to update the docs to replace the deprecated method calls.

**Comment by morellik** [comment]
Thank you very much!!!

It's works as expected.

But how can I add rows to a new DataFrame?

If I use
 df.vConcat(PDBDataFrame
                    .rows($int("atomNumber").mapBoolVal(a -> a == Integer.parseInt(l.get(5))))
                    .select(...));

I receive java.lang.NullPointerException: Cannot invoke "org.dflib.DataFrame.vConcat(org.dflib.DataFrame[])" because "this.df" is null] with root cause

If I use df = PDBDataFrame
                    .rows($int("atomNumber").mapBoolVal(a -> a == Integer.parseInt(l.get(5))))
                    .select(...));

I've only the last row.


Another question, thank you for your patience, when I change a field value, is there a way to select the field and set the value leaving the other unchanged?

I've a lot of fields and for the moment I use:

            PDBDataFrame
                    .rows($int("atomNumber").mapBoolVal(a -> a == Integer.parseInt(l.get(5))))
                    .select(
                            $col("record"),
                            $col("atomNumber"),
                            $col("atomName"),
                            $col("residueName"),
                            $col("chain"),
                            $col("residueNumber"),
                            $col("x"),
                            $col("y"),
                            $col("z"),
                            $col("occupancy"),
                            $val("40.00"),
                            $col("segID"),
                            $col("elementType")
                    ));

depending on the value of a particular field I have to repeat the loop at least 5 times and it becomes tedious to repeat the sequence of fields all the times.
It would take a way to be able to assign a given field the new value while leaving the others unchanged.
Is this possible?

**Comment by andrus** [comment]
> I receive java.lang.NullPointerException: Cannot invoke "org.dflib.DataFrame.vConcat(org.dflib.DataFrame[])" because "this.df" is null] with root cause

This has to be something in your code.

>  It would take a way to be able to assign a given field the new value while leaving the others unchanged.

The details may vary, but try using `map` instead of `select` . `map` allows to change data, and merge it with the source DataFrame

**Comment by morellik** [comment]
Thanks. I'll try to use map with RowMapper.

Now I've a DataFrame with a content coming from a file. How can I create another DataFrame giving the rows from the first DataFrame when a condition is true?

         List<DataFrame> df1 = new ArrayList<>();
         DataFrame df = DataFrame.empty();

        for (List<String> l : metal2donors.get(metal)) {
               
            RowMapper mapper = (from, to) -> {
                    from.copy(to);
                    to.set("tempFactor", "40.00");
            };
            df1.add(PDBDataFrame
                    .rows($int("atomNumber").mapBoolVal(a -> a == Integer.parseInt(l.get(5))))
                    .map(mapper
                    ));
        }

        df = df.vConcat(df1.toArray(new DataFrame[0]));
        String table = Printers.tabular.toString(df);
        System.out.println(table);

clearly this code does not work.  df1 hasn't rows.
But at the end, df1 would have all the columns of PDBDataFrame with one column modified. Is it possible?

**Comment by morellik** [comment]
Sorry, but it is more difficult for me to understand. I've tried the following code:

    List<DataFrame> df1 = new ArrayList<>();
        DataFrame df = DataFrame.empty();
        for (List<String> l : metal2donors.get(metal)) {
                //df1.add(PDBDataFrame.rows(r -> r.get("atomNumber", Integer.class).equals(Integer.parseInt(l.get(8)))).select());
            RowMapper mapper = (from, to) -> {
                    from.copy(to);
                    to.set("tempFactor", "40.00");
            };
            /*df1.add(PDBDataFrame
                    .rows($int("atomNumber").mapBoolVal(a -> a == Integer.parseInt(l.get(5))))
                    .map(mapper
                    ));*/

            df1.add(PDBDataFrame
                    .rows($int("atomNumber").mapBoolVal(a -> a == Integer.parseInt(l.get(5))))
                    .map(
                            $col("record"),
                            $col("atomNumber"),
                            $col("atomName"),
                            $col("residueName"),
                            $col("chain"),
                            $col("residueNumber"),
                            $col("x"),
                            $col("y"),
                            $col("z"),
                            $col("occupancy"),
                            $val("40.00"),
                            $col("segID"),
                            $col("elementType")
                    ));

        }

        df = df.vConcat(df1.toArray(new DataFrame[0]));
        String table = Printers.tabular.toString(df);
        System.out.println(table);
        return df;

But at the end, df is empty :-((
I'm going crazy....

**Comment by andrus** [comment]
I think I now better understand what you are trying to do. Actually, `select(...)` should work in your case. Here is a simple example that selects the rows where "b" is 5, and replaces "a" with "X":

The initial DataFrame:

```
a b
- -
a 5
b 6
c 5
```

The transform:
```java
df.rows($int("b").eq(5))
    .select($val("X"), $col("b"));
```

The result:
```
a b
- -
X 5
X 5
```

_FWIW, `map` can potentially allow you to avoid the loop and vConcat. But let's make sure you current implementation works first_
